### PR TITLE
Adopt a Code of Conduct for Dolphin.

### PR DESCRIPTION
**Pull request by delroth**
[Rendered](https://github.com/delroth/dolphin/blob/coc/CODE_OF_CONDUCT.md)

This stems from the current discussions around having an official Dolphin Discord. We're going to have more users and developers interacting over there than we've had before, and given the general state of gaming communities on the internet I'd like to make it clear what's acceptable and not acceptable for us. If we define clearer rules for Discord, I don't really see any reason why not apply them globally to anywhere humans interact for the project: GitHub, bug tracker, IRC, forums, etc.

This Code of Conduct is derived from https://asahilinux.org/code-of-conduct/ which itself has a long lineage in the open source community (used by Django and Ceph, two other major open source projects of comparable size to ours). I've made roughly 0 adjustments to the CoC text compared to the Asahi version at this point. This PR is an RFC so we can discuss if we want to change anything before applying this to our project.

Two main points that I think need to be addressed before merge:

- [x] Who should be triaging reports of misconducts sent to conduct@dolphin-emu.org? We don't officially have any kind of "project leads" for Dolphin, and I think we'd want to have a decent amount of diversity of viewpoints in this group. I think starting with 3 trusted long-term project members would be fine for now, hopefully this is a very low-volume source of work :-)
- [x] I'm not super happy about the "Consequences" section as it is -- I like its general focus on allowing people to make mistakes as long as they stop misbehaving once notified, but I suspect that due to the nature of its userbase Dolphin might have to deal with on average more "malicious" / "deliberately harmful" people than this was designed for. Technically the wording allows skipping directly to the 3rd enforcement action, but I think this should maybe be made more explicit (-> we won't give you the benefit of the doubt if your intent seems unmistakenly malicious).

### CODE REVIEW

**Comment by delroth** [comment]
One option raised by @K0bin on IRC might be to just get rid of the "Consequences" section altogether, or at least greatly simplify it. I think that might be a good idea, especially since enforcement of the rules will happen in a very disparate fashion and with different groups of moderators depending on the communications platform.

I think I wouldn't completely get rid of it though and still have it describe our general intent: assume good faith when it seems plausible, and allow people to make mistakes as long as there's no pattern of undesirable behavior.

**Comment by marcan** [comment]
Keep in mind that the purpose of CoCs (IMO) is to outline the general approach of a community, signal that it cares about these things, and be a set of guidelines, rather than to serve as strictly interpreted laws of the land. So there is always room for discretion in how it is applied, especially in egregious cases.

That's not to say you shouldn't change the "Consequences" section, that's perfectly fine, just that you shouldn't worry *too* much about having a very strictly accurate set of rules because the people who are going to try to rules-lawyer you after you ban them are the people you want to stay banned anyway.

**Comment by delroth** [comment]
> That's not to say you shouldn't change the "Consequences" section, that's perfectly fine, just that you shouldn't worry _too_ much about having a very strictly accurate set of rules because the people who are going to try to rules-lawyer you after you ban them are the people you want to stay banned anyway.

I agree, but that's exactly why the current wording of that section feels weird to me -- I find that it reads in a very precise/procedural way when in practice, as you mention, it's the general intent and guidelines that matter. Which is why I like the suggestion from my previous comment of rewriting the section to focus on the intent and the guidelines we use to decide on moderation actions.

**Comment by delroth** [comment]
> (I suspect that there might be communities that can afford funneling all the CoC-relevant moderation decisions through a small group of people that can guarantee consistency and a uniform moderation process. In which case a wording like the current one might make more sense? But I don't think Dolphin can ever work like this due to the shear amount of moderation we have to do for user-facing comms.)

Another difference I can see is that projects like Django / Ceph have to interact with a more corporate side in their community, and someone being blocked from contributing to the project could directly lead to very negative consequences for that person (like, I can see someone losing their job over it?). Under these constraints it probably makes more sense to have a much clearer and precise list of enforcement actions and the conditions under which they're used. But that's not something that I think applies to Dolphin, and that difference might justify approaching the wording differently.

**Comment by JMC47** [comment]
I did a quick read through it.  It seems reasonable enough, but the consequences section can probably be removed or something more akin to perpetually problematic people being removed from the community.

**Comment by lioncash** [comment]
I am 100% in favor with adopting a code of conduct. It's up to maintainers to get rid of members that would be ultimately harmful to efforts at collaboration. And I think the overall CoC is very amenable to regulating the community in a fair way,

I do think a consequences section should exist to outline, that, should the majority or leadership portion of a community think one is a problematic individual for a whole community (in the sense of dragging it down), they should be allowed to be removed without warning (there's no reason to keep someone problematic around for diplomatic reasons, because the situation is not a democratic one, but an analytical one by existing enforcing members), but I also don't care to argue about that, so have at it).

A lot of communities I've been part of have kept antagonistic or problematic entities along the guidelines of "well, they didn't *technically* break anything" to the detriment of everyone around them, so I'd like to be explicit in saying "no, if you're a shithead, you're gone. Total moderator discretion".

I'd probably be able to smooth this out with more time, but it's also almost 2:40 in the morning, so alas, bear with me.

**Comment by JMC47** [comment]
That's fine with me.  I'm not exactly an expert opinion on this and my opinion isn't that strong on it.

**Comment by MayImilae** [comment]
I think we as a project are well behind having a code of conduct. Glad to see we're finally getting one!

I concur that the consequences section should exist, but in a form that gives more freedom to allow trusted members to act quickly as needed. While not directly open source CoC experience, I have a lot of experience with moderation of various communities. A number of those communities had very complex rules of do X get Y punishment and frankly... it never worked well. Having loose internal guidelines for how to act in typical circumstances can be useful, but it should never be rigidly coded as "rules".

The way Asashi's consequences section is designed, it's meant to be a buffer to help prevent those in power from abusing newcomers. Though admirable, I think it is misguided. If someone high up in a project wants to be an ass to newcomers, they'll dance around rigid rules like they were nothing. I've seen it happen. In my experience, this kind of approach won't stop a bad actor from harming newcomers, and it limits down the responsiveness of those with good intentions trying to deal with immediate threats. ...also Asashi's consequences section totally ignores the existence of spam, trolling, hate attacks, and just, the worst of the internet that requires VERY rapid and aggressive responses.

So imo, the consequences section, if we have one, should just say that trusted members will act as needed, and if there is a grievance, here's an email. This lets us act fast and gives trusted members the freedom to take care of external threats rapidly. Then the email is for internal CoC issues such as abuses and stuff from those on the inside.

As for how the consequences section should be worded, [Contributor Covenant](https://www.contributor-covenant.org/version/1/4/code-of-conduct/) is a good reference.

**Comment by MayImilae** [comment]
As an aside, The CoC that Asashi's Code of Conduct is based on, [django's](https://www.djangoproject.com/conduct/), does not have a consequences section at all - that was added by Ceph.io.

django's doesn't have the consequences section and just has an email for reporting any abuses. IMO, that's a great route to go.

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+contributors and users from around the globe. We find the contributions,
+collaborations, and mentorships within our community to be the essential
+lifeblood of our project and appreciate the efforts of those who participate to
+nurture and grow those, and all other aspects of our community.
+
+However, when a large and sufficiently diverse group of people work together,
+there are often cultural, communication, and compatibility issues. In order to
+minimize conflict, and provide a framework for resolution, we have a brief code
+of conduct that we ask all participants in the Dolphin community adhere to.
+These rules should apply to everyone, regardless of station within the
```
Comment (JosJuice) [inline CODE_OF_CONDUCT.md:14]: This is just a small nitpick, but I don't understand what "station" means in this context.

Comment (merryhime) [reply CODE_OF_CONDUCT.md:14]: suggest `s/station/status/`

**Comment by mbc07** [comment]
Just read the current version and all topics seems reasonable to me, nothing to add, LGTM...

**Comment by delroth** [comment]
@lioncash @JosJuice @MayImilae PTAL.

We still need to figure out who should be on conduct@dolphin-emu.org. Are any of you interested?

(I would volunteer but also realize that I'm uhhhh not always the most well behaving person, as much as I try.)

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (delroth) [reply CODE_OF_CONDUCT.md]: For reference that would be definition 6 on https://en.wiktionary.org/wiki/station#Noun, I agree it's not super commonly used in that meaning so I replaced it with @merryhime's suggestion.

**Comment by lioncash** [comment]
I would be fine with being on it, but I'd also prefer if a second person were also on it as well, since I can be pretty busy some parts of the year and might not be able to reply to things as quick as I'd like.

**Comment by MayImilae** [comment]
We should have a fair number of people on it so I'll volunteer as well.

**Comment by delroth** [comment]
For the sake of unblocking this I will start with the following list on conduct@dolphin-emu.org:

- @delroth (me)
- @lioncash 
- @MayImilae 

Subject to change if someone else volunteers.

Are there any remaining review comments from anyone? I'll leave this open for another day and merge tomorrow otherwise.

**Pull request merged**